### PR TITLE
Demo testing pick and place

### PR DESCRIPTION
**Pull request by MDecarabas**

### CODE REVIEW

File: crane_x7_tests/crane_x7_tests/__init__.py
Comment (fabid) [inline crane_x7_tests/crane_x7_tests/__init__.py:1]: Is this file needed for some reason?

File: crane_x7_tests/test/test_pep257.py
Comment (fabid) [inline crane_x7_tests/test/test_pep257.py:1]: question, are those automatically generated tests  (flake8, pep257, copyright)  passing?

File: crane_x7_tests/setup.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        ('share/ament_index/resource_index/packages',
+            ['resource/' + package_name]),
+        ('share/' + package_name, ['package.xml']),
+    ],
+    install_requires=['setuptools'],
+    zip_safe=True,
+    maintainer='decarabas',
+    maintainer_email='codrea.eric@gmail.com',
+    description='TODO: Package description',
+    license='TODO: License declaration',
```
Comment (fabid) [inline crane_x7_tests/setup.py]: what is the license used in the rest of the Crane package?

File: crane_x7_tests/resource/crane_x7_tests
Comment (fabid) [inline crane_x7_tests/resource/crane_x7_tests:1]: does this empty file has a purpose?

File: crane_x7_tests/package.xml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  <name>crane_x7_tests</name>
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
+  <maintainer email="codrea.eric@gmail.com">decarabas</maintainer>
+  <license>TODO: License declaration</license>
+
+  <test_depend>ament_copyright</test_depend>
+  <test_depend>ament_flake8</test_depend>
+  <test_depend>ament_pep257</test_depend>
+  <test_depend>python3-pytest</test_depend>
```
Comment (fabid) [inline crane_x7_tests/package.xml:13]: I suppose the test aso depends on `crane_x7_examples`

File: crane_x7_tests/launch/pick_and_place.launch.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            "use_sim_time": "true", 
+            "example": "pick_and_place"
+        }.items(),
+    )
+
+    
+    return LaunchDescription(
+        [
+            sim,
+            TimerAction(period=15.0, actions=[demo]),  # Need to optimize period, if under 10 seconds will not run
```
Comment (fabid) [inline crane_x7_tests/launch/pick_and_place.launch.py]: Thank you, nice to have a comment for those.

Indeed, we may want a better solution at some point, but I think this is good as is for now

File: crane_x7_tests/launch/pick_and_place.launch.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        position = matches[-2]
+
+        # Extract the individual numbers from the matched string
+        numbers = re.findall(r"[\d.-]+", position)
+        
+        if len(numbers) >= 3:
+            x_variable = float(numbers[0])
+            y_variable = float(numbers[1])
+            z_variable = float(numbers[2])
+            print(f"x box coordinate: {x_variable}")
```
Comment (fabid) [inline crane_x7_tests/launch/pick_and_place.launch.py]: we should get rid of the print statements here

File: crane_x7_tests/launch/pick_and_place.launch.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+def get_model_location(model_name):
+    """
+    Function to get the location of a model
+    """
+
+    command = ["ign", "model", "-m", model_name, "-p"]
+    result = subprocess.run(command, check=True, text=True, stdout=subprocess.PIPE)
+    result_output = result.stdout
+
+    print(result_output)
```
Comment (fabid) [inline crane_x7_tests/launch/pick_and_place.launch.py]: print statement

File: crane_x7_tests/launch/pick_and_place.launch.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    return LaunchDescription(
+        [
+            sim,
+            TimerAction(period=15.0, actions=[demo]),  # Need to optimize period, if under 10 seconds will not run
+            launch_testing.actions.ReadyToTest(),
+        ]
+    ), {"demo": demo, "sim":sim}
+
+
+def get_model_location(model_name):
```
Comment (fabid) [inline crane_x7_tests/launch/pick_and_place.launch.py:71]: Do you remember when you first worked on that? I think at the beginning wer were planning to use some service, however it was not working with gazebo classic if I recall. Since we have ignition here, is there a better solution?

If not it could be interesting to put that in some helper library at some point (either a standalone package, or part of the artefacts client)

File: crane_x7_tests/launch/pick_and_place.launch.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        model_name = "wood_cube_5cm"
+
+        (
+            x_variable,
+            y_variable,
+            z_variable,
+        ) = get_model_location(model_name)
+
+        # proc_output.assertWaitFor("Plan and Execute request complete!", timeout=180)
+        sleep(40)
```
Comment (fabid) [inline crane_x7_tests/launch/pick_and_place.launch.py]: Do you think there could be a better way than a sleep here? If we need to add  a single line ros log statement to the  PickAndPlace example code to allow for an assertWaitFor, I think this is acceptable

File: crane_x7_tests/launch/pick_and_place.launch.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+        (
+            x_new_variable,
+            y_new_variable,
+            z_new_variable,
+        ) = get_model_location(model_name)
+
+        self.assertNotEqual(x_variable, x_new_variable, decimal)
```
Comment (fabid) [inline crane_x7_tests/launch/pick_and_place.launch.py]: Since we are not doing a "push" test, it may be better to test that the cube ends up at/near the desired place location.  Because if for some reason, the cube gets picked up but falls from the gripper on the way, we want that to be marked as false.

Ideally, we would like  import the place location from the example (this way if the coordinates are changed in the example, the test does not break). But this may not be possible/easy. In that case, you can log the final position once, and use that for the test (maybe rounded) and compare with some reasoble margins ( 1cm?)

**Review by fabid — COMMENTED** [review]
Thank you for the PR. I like that it is non invasive and overall simple enough.

Some comments for discussion

File: crane_x7_tests/crane_x7_tests/__init__.py
Comment (MDecarabas) [reply crane_x7_tests/crane_x7_tests/__init__.py:1]: This is an automatic file created by `ros2 pkg create --build-type ament_python <package_name>` when creating a ros2 package

File: crane_x7_tests/launch/pick_and_place.launch.py (hunk 7 above)
Comment (MDecarabas) [reply crane_x7_tests/launch/pick_and_place.launch.py]: If that is acceptable that is definitely the way to go. As you pointed out the only reason this sleep exists is because I had no log to link to so that I could check that the pick and place motion had finished executing

File: crane_x7_tests/resource/crane_x7_tests
Comment (MDecarabas) [reply crane_x7_tests/resource/crane_x7_tests:1]: This is an automatic file created by ros2 pkg create --build-type ament_python <package_name> when creating a ros2 package

File: crane_x7_tests/launch/pick_and_place.launch.py (hunk 6 above)
Comment (MDecarabas) [reply crane_x7_tests/launch/pick_and_place.launch.py:71]: As we discussed yesterday despite this not being the cleaner way to solve the issue, it is less invasive in relation to changes we would make to the code. I agree that a really good idea would be to have an "artefacts_sim_bridge" package that would allow for easy access to gz-classic and ignition variables/models/etc...

File: crane_x7_tests/launch/pick_and_place.launch.py (hunk 8 above)
Comment (MDecarabas) [reply crane_x7_tests/launch/pick_and_place.launch.py]: Pick and place test is done. WIP: getting the final values without hard coding them in the test

File: crane_x7_tests/resource/crane_x7_tests
Comment (MDecarabas) [reply crane_x7_tests/resource/crane_x7_tests:1]: https://answers.ros.org/question/386756/ros2-design-resource-directory/
Reason for why this file exists. Tl;dr: It is required for ROS2 to acknowledge the existence of a package

File: crane_x7_tests/setup.py (hunk 1 above)
Comment (MDecarabas) [reply crane_x7_tests/setup.py]: Apache license 2.0. I added the same one to ours so we are congruent with what they do

File: crane_x7_tests/launch/pick_and_place.launch.py (hunk 8 above)
Comment (fabid) [reply crane_x7_tests/launch/pick_and_place.launch.py]: Hard coding the values is perfectly acceptable if the alternative is too complicated

File: crane_x7_tests/launch/pick_and_place.launch.py (hunk 6 above)
Comment (fabid) [reply crane_x7_tests/launch/pick_and_place.launch.py:71]: as discussed, the ideal solution would be using ros params, but we can skip for now
